### ISSUE TITLE
improve memory usage

### ISSUE THREAD

**Issue by veggiesaurus** [idia-astro/cpp_hdf5converter#4]
We need to figure out how to limit memory usage. Currently, it reserves space in memory equal to twice the size of the 3D cube. With some optimisations as to how we read things, we could probably reduce this to 1x the size of the 3D cube plus some constant value

In future, we can try to reduce this further by using a tiling approach.

**Comment by confluence** [comment]
Further issues that I found after fixing the memory allocation reporting:

* Z stats use a huge amount of space, which can greatly exceed all the other memory use combined if the number of channels is low. This is because we accumulate stats in multiple 64-bit vectors, and for Z stats the size of each vector is the size of the XY dimensions of the image.

* We unnecessarily store stats for the whole image at a time. Stats never cross multiple stokes, so we could store only one stokes at a time.

* Mipmaps store a fair amount of memory because we use doubles to calculate them and additionally store the number of not-NAN values that contribute to every downsampled value (so that we divide by the correct number to perform the downsampling). We could perhaps just use floats for this.

Possible solutions:

* Stop storing stats for all stokes. This is straightforward.

* In the slow converter we could probably move the Z stats calculation to the loop which performs tiled swizzling. In the fast converter we could possibly calculate them in a separate pass over the rotated dataset.

* Maybe use floats to downsample the mips, and consider an alternative algorithm which performs the division immediately and avoids storing the not-NAN count. This may not be a major issue because in the fast algorithm the size is still consistently below the size of the main dataset, and we deallocate the rotated dataset before allocating the mipmaps, and in the slow converter we only store mipmaps for one channel at a time.

**Comment by confluence** [comment]
Some of this has been resolved in the current development version.

A major outstanding issue, which should be resolved in the upcoming work to produce an optimised version which fits within a specified memory limit, is the allocation of tiled slices in the slow converter (which always uses the full tile size, even if the dataset is smaller than a tile, and always uses the full depth even if it can't fit in memory because the image is very deep).

**Comment by confluence** [comment]
This is a very broad issue, some of which has already been resolved. Should we close this and also #25 and open a new issue specifically for creating a "smart" converter mode (which splits the work by channel / tile like the slow mode but also processes the work units in parallel)?

**Comment by veggiesaurus** [comment]
good idea

**Comment by confluence** [comment]
As discussed, this issue has been superseded by #44.
